### PR TITLE
Servlet request wrapper

### LINKED ISSUE #44: HttpServletRequestWrapper needed
Implement methods:
   public String getRemoteUser()
   public java.security.Principal getUserPrincipal()
   public boolean isUserInRole( String role )
from javax.servlet.http.HttpServletRequestWrapper

### PR DESCRIPTION
**Pull request by victornoel**
This closes #44.

@mores what this what you had in mind? :)

### CODE REVIEW

File: src/main/java/org/pac4j/j2e/util/Pac4JHttpServletRequestWrapper.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.Collection;
+import java.util.Optional;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletRequestWrapper;
+
+import org.pac4j.core.profile.CommonProfile;
+import org.pac4j.core.profile.ProfileHelper;
+
+public class Pac4JHttpServletRequestWrapper extends HttpServletRequestWrapper {
```
Comment (leleuj) [inline src/main/java/org/pac4j/j2e/util/Pac4JHttpServletRequestWrapper.java]: Javadoc expected.

File: src/main/java/org/pac4j/j2e/util/Pac4JHttpServletRequestWrapper.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    public Pac4JHttpServletRequestWrapper(HttpServletRequest request, Collection<CommonProfile> profiles) {
+        super(request);
+        this.profiles = profiles;
+    }
+
+    @Override
+    public String getRemoteUser() {
+        // TODO or id?
+        return getProfile().map(CommonProfile::getUsername).orElse(null);
```
Comment (leleuj) [inline src/main/java/org/pac4j/j2e/util/Pac4JHttpServletRequestWrapper.java]: Yes, it would use the identifier as a fallback.

File: pom.xml [hunk 3]
```diff
@@ -53,7 +53,7 @@
     </developers>
 
     <properties>
-        <pac4j.version>2.1.0</pac4j.version>
+        <pac4j.version>3.0.0-SNAPSHOT</pac4j.version>
```
Comment (leleuj) [inline pom.xml]: As we upgrade to a major pac4j version, it should be a major j2e-pac4j version (4.0.0-SNAPSHOT).

Comment (victornoel) [reply pom.xml]: yep, noted

Comment (victornoel) [reply pom.xml]: (it's too bad j2e-pac4j is always one major version in advance from pac4j : )

**Comment by victornoel** [comment]
I've pushed some changes, should be good now

**Comment by victornoel** [comment]
@leleuj the version of pac4j-core that travis is downloading is from september, I think there is something wrong either with pac4j deployment or travis configuration for this project...

**Comment by victornoel** [comment]
@leleuj apparently the deployment is not done because of a bug found by findbug: https://travis-ci.org/pac4j/pac4j#L5506

**Comment by leleuj** [comment]
Indeed, the oldest pac4j v3.0.0-SNAPSHOT artifact is from September, the 12th: https://oss.sonatype.org/content/repositories/snapshots/org/pac4j/pac4j-core/3.0.0-SNAPSHOT/

I will investigate...

**Comment by leleuj** [comment]
It should be ok now...

**Comment by victornoel** [comment]
@leleuj great thanks, then I think you can finalize the review on this :)

**Comment by leleuj** [comment]
@victornoel Excellent. Two more points:

1) People should be aware of that capability. I think you should add a few words about that in the "5) Get the user profile (ProfileManager)" section of the README

2) We need to update the demo later on, though I can handle that myself.

**Comment by victornoel** [comment]
I added some basic documentation and improved the code a bit, if you are good with this, then it can be merged :)

File: src/main/java/org/pac4j/j2e/util/Pac4JHttpServletRequestWrapper.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    private Collection<CommonProfile> profiles;
+
+    public Pac4JHttpServletRequestWrapper(HttpServletRequest request, Collection<CommonProfile> profiles) {
+        super(request);
+        this.profiles = profiles;
+    }
+
+    @Override
+    public String getRemoteUser() {
+        return getPrincipal().map(p -> p.getName()).orElse(null);
```
Comment (leleuj) [inline src/main/java/org/pac4j/j2e/util/Pac4JHttpServletRequestWrapper.java]: You removed the identifier as a fallback if the username is blank? I thought it was better the other way...

Comment (victornoel) [reply src/main/java/org/pac4j/j2e/util/Pac4JHttpServletRequestWrapper.java]: It's because our implementation of Principal already those this for us :)

Comment (leleuj) [reply src/main/java/org/pac4j/j2e/util/Pac4JHttpServletRequestWrapper.java]: You're right. Good point.

**Pull request merged**